pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -545,6 +557,15 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 		}
 	}
 
+	// kubelet restart flow.
+	if !m.sourcesReady.AllReady() {
+		// if the container is reported running and it had a device assigned, things are already fine and we can just bail out
+		if needed == 0 && m.checkContainerAlreadyRunning(podUID, contName) {
```

Comment:
Question: Is the `needed == 0` check relevant here? We already error out above if this is not the case. Furthermore, would moving the
```
	if needed == 0 {
		klog.V(3).InfoS("no devices needed, nothing to do", "deviceNumber", needed, "resourceName", resource, "podUID", string(podUID), "containerName", contName)
		// No change, no work.
		return nil, nil
	}
``` 
below earlier in this function not achieve the same result, or are the additional checks required in any case?

Reply:
I was thinking to a case like pod mutated (resource request changed, e.g. adding a device) while kubelet is down (or restarting and not yet recovered from the apiserver, which is functionally the same case) but thinking more about it should not be possible to mutate the request without deleting/recreating the pod, not even with Vertical Pod Autoscaling.
I'll doublecheck this and simplify the flow if possible. Thanks for highlighting this.

Reply:
I believe the check for `needed == 0` below can be pulled above this (removing the need to check it here).

Reply:
We need to tread carefully here because the check with `needed == 0` was a source of troubles in the recent past. I'm adding better comments to explain the flow here, I think they are warranted. Furthermore, in the end we _can_ pull this check above indeed if we check `requested` (= the original allocation request) instead of `needed` (= the fixed allocation request considering the already allocated devices from the checkpoint file).

I think we can add a check here: https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.4/pkg/kubelet/cm/devicemanager/manager.go#L772 and shortcircuit earlier, avoiding `devicesToAllocate an all its complexity entirely. I'm however deferring this work to another PR to avoid further complicate this fix.